mountpoint-s3/src/fs.rs

```diff
@@ -464,26 +471,20 @@ where
             }
         };
         let mut request = match &handle.typ {
-            FileHandleType::Write { parts, .. } => parts.lock().await,
+            FileHandleType::Write { request, .. } => request.lock().await,
             FileHandleType::Read { .. } => return Err(libc::EBADF),
         };
 
-        let next_offset = request.iter().map(|p| p.len()).sum::<usize>();
-        if offset != next_offset as i64 {
-            error!("out of order write; expected offset {next_offset} but got {offset}");
-            return Err(libc::EINVAL);
-        }
-
-        // If we'd go over the size limit, fail the entire write rather than short-writing
-        if next_offset + data.len() > MAX_OBJECT_SIZE {
-            error!("object too large");
-            return Err(libc::EFBIG);
+        match request.write(offset, data).await {
+            Ok(()) => {
+                let len = data.len();
+                Ok(len as u32)
+            }
+            Err(e) => {
+                error!("write failed: {:?}", e);
+                Err(e.into())
```

Comment:
Hmm, I'm wondering what we should do with the upload if a write fails like this. Is it better to persist the object, knowing that we've dropped some writes and so it's not the object the customer expected? Or should we abandon the upload completely?

Reply:
Good question. My first answer would be that we should abort, but I have not thought it through:
* we can abort the upload, but what about the file handle? just keep failing on subsequent writes?
* is there a difference between writes erroring out immediately (e.g. out-of-order) and partial writes?
* where do write checksums fit in? 

Worth raising a separate issue?

Reply:
Yeah, let's track it separately.